objects/src/accounts/delta/vault.rs

```diff
@@ -37,6 +37,28 @@ impl AccountVaultDelta {
         }
     }
 
+    /// Merges another delta into this one, overwriting any existing values.
+    pub fn merge(self, other: Self) -> Self {
+        // Collec the assets into a hashmap where true indicates added, and false removed.
+        // This lets us simplify the addition/removal dance we would otherwise have to go through.
+        //
+        // Ordering matters. Since later items will overwrite earlier ones we need to begin
+        // with self, then other.
+        let assets = self
+            .added_assets
+            .into_iter()
+            .map(|asset| (asset, true))
+            .chain(self.removed_assets.into_iter().map(|asset| (asset, true)))
+            .chain(other.added_assets.into_iter().map(|asset| (asset, false)))
+            .chain(other.removed_assets.into_iter().map(|asset| (asset, false)))
+            .collect::<BTreeMap<_, _>>();
```

Comment:
We should update the comment to mention `BTreeMap` instead of hashmap, but also - I'm not sure this logic is correct. Specifically, I'm not sure it will merge fungible assets correctly.

For example, if `self` has added asset `(account_id1, 100)` and `other` has added asset `(account_id1, 200)`, as far as I can tell, these would not be merged into a single asset `(account_id1, 300)`.

I think we probably should process fungible and non-fungible assets separately here. For non-fungible assets, something like the above logic should work. For fungible assets, we probably need something like `BTreeMap<AccountId, amount>` where amount could be either positive or negative (i.e., we could use `i64`).